File: algo/4. Median of Two Sorted Arrays.js
// @9 Array

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function(nums1, nums2) {
    var helper = function(nums1, index1, nums2, index2, k) {
        // if the answer is in nums2
        if (index1 > nums1.length -1) {
            return nums2[index2+k-1];
        }
        // if the answer is in nums1
        if (index2 > nums2.length-1) {
            return nums1[index1+k-1];
        }
        // if k has been decreased to 1
        if (k==1) {
            return Math.min(nums1[index1], nums2[index2]);
        }
        // add Infinity to the bottom of two arrays
        var mid1 = Infinity, mid2 = Infinity;
        // compare two k/2 indexed values to decide
        // which k/2 useless values need to remove
        if (index1+Math.floor(k/2)-1 < nums1.length) {
            mid1 = nums1[index1+Math.floor(k/2)-1];
        }
        if (index2+Math.floor(k/2)-1 < nums2.length) {
            mid2 = nums2[index2+Math.floor(k/2)-1];
        }
        if (mid1 < mid2) {
            return helper(nums1, index1+Math.floor(k/2), nums2, index2, k-Math.floor(k/2));
        } else {
            return helper(nums1, index1, nums2, index2+Math.floor(k/2), k-Math.floor(k/2));
        }
    };
    // consider both scienarios, even length or odd length
    var mid1 = Math.floor((nums1.length+nums2.length+1)/2);
    var mid2 = Math.floor((nums1.length+nums2.length+2)/2);
    if(mid1 == mid2) {
        return helper(nums1, 0, nums2, 0, mid1);
    }
    return (helper(nums1, 0, nums2, 0, mid1) + helper(nums1, 0, nums2, 0, mid2)) / 2.0;
};

File: round2/4. Median of Two Sorted Arrays.cs
public class Solution {
    private int Helper(int[] nums1, int index1, int[] nums2, int index2, int k) {
        if (index1 > nums1.Length - 1) {
            return nums2[k+index2-1];
        }
        if (index2 > nums2.Length - 1) {
            return nums1[k+index1-1];
        }
        if (k == 1) {
            return nums1[index1] < nums2[index2] ? nums1[index1] : nums2[index2];
        }
        int mid1 = int.MaxValue, mid2 = int.MaxValue;
        if (index1+(int) k/2-1 < nums1.Length) {
            mid1 = nums1[index1+(int) k/2-1];
        }
        if (index2+(int) k/2-1 < nums2.Length) {
            mid2 = nums2[index2+(int) k/2-1];
        }
        if (mid1 < mid2) {
            return Helper(nums1, index1+(int) k/2, nums2, index2, k-(int) k/2);
        } else {
            return Helper(nums1, index1, nums2, index2+(int) k/2, k-(int) k/2);
        }
    }
    
    public double FindMedianSortedArrays(int[] nums1, int[] nums2) {
        int mid1 = (int) (nums1.Length + nums2.Length + 1) / 2;
        int mid2 = (int) (nums1.Length + nums2.Length + 2) / 2;
        if (mid1 == mid2) {
            return Helper(nums1, 0, nums2, 0, mid1);
        } else {
            return (Helper(nums1, 0, nums2, 0, mid1) + Helper(nums1, 0, nums2, 0, mid2)) / 2.0;
        }
    }
}